modules/rostests/kmtests/ntos_io/IoDeviceInterface.c

```diff
@@ -1,125 +1,214 @@
+
+        if (NT_SUCCESS(Status))
+        {
+            trace("ZwCreateKey(): successfully created subkey: %d %p\n", n, DeviceInterfaceSubKey);
+
+            ZwDeleteKey(DeviceInterfaceSubKey);
+            ZwClose(DeviceInterfaceSubKey);
+        }
+        else
+        {
+            trace("ZwCreateKey(): failed to create a subkey: %d 0x%x\n", n, Status);
+        }
+
+        ZwClose(DeviceInterfaceKey);
+    }
 }
 
 static
 VOID
-Test_IoRegisterDeviceInterface(VOID)
+Test_IoGetDeviceInterfaceAlias(PWSTR SymbolicLink)
 {
-    GUID Guid = {0x378de44c, 0x56ef, 0x11d1, {0xbc, 0x8c, 0x00, 0xa0, 0xc9, 0x14, 0x05, 0xdd}};
-    DEVICE_OBJECT DeviceObject;
-    EXTENDED_DEVOBJ_EXTENSION DeviceObjectExtension;
-    DEVICE_NODE DeviceNode;
     UNICODE_STRING SymbolicLinkName;
-    NTSTATUS Status;
 
-    RtlInitUnicodeString(&SymbolicLinkName, L"");
+    RtlInitUnicodeString(&SymbolicLinkName, SymbolicLink);
+
+    for (size_t n = 0; n < RTL_NUMBER_OF(Types); ++n)
+    {
+        UNICODE_STRING AliasSymbolicLinkName;
+        NTSTATUS Status = IoGetDeviceInterfaceAlias(&SymbolicLinkName, Types[n], &AliasSymbolicLinkName);
+
+        if (NT_SUCCESS(Status))
+        {
+            trace("IoGetDeviceInterfaceAlias(): success: %d %wZ\n", n, &AliasSymbolicLinkName);
+
+            RtlFreeUnicodeString(&AliasSymbolicLinkName);
+        }
+        else
+        {
+            trace("IoGetDeviceInterfaceAlias(): fail: %d 0x%x\n", n, Status);
+        }
+    }
+}
+
+static
+VOID
+Test_IoSetDeviceInterfaceState(PWSTR SymbolicLink)
+{
+    UNICODE_STRING SymbolicLinkName;
+
+    RtlInitUnicodeString(&SymbolicLinkName, SymbolicLink);
+
+    for (size_t n = 0; n < RTL_NUMBER_OF(Types); ++n)
+    {
+        NTSTATUS Status = IoSetDeviceInterfaceState(&SymbolicLinkName, TRUE);
+
+        if (NT_SUCCESS(Status))
+        {
+            trace("IoSetDeviceInterfaceState(): successfully enabled interface: %d %wZ\n", n, &SymbolicLinkName);
+        }
+        else
+        {
+            trace("IoSetDeviceInterfaceState(): failed to enable interface: %d 0x%x\n", n, Status);
+        }
+    }
+}
+
+static
+VOID
+Test_IoGetDeviceInterfaces(const GUID* guid)
+{
+    NTSTATUS Status;
+    PWSTR SymbolicLinkList, SymbolicLinkListPtr;
+    UNICODE_STRING GuidString;
 
-    // Prepare our surrogate of a Device Object
-    DeviceObject.DeviceObjectExtension = (PDEVOBJ_EXTENSION)&DeviceObjectExtension;
+    RtlStringFromGUID(guid, &GuidString);
 
-    // 1. DeviceNode = NULL
-    DeviceObjectExtension.DeviceNode = NULL;
-    Status = IoRegisterDeviceInterface(&DeviceObject, &Guid, NULL,
-        &SymbolicLinkName);
+    Status = IoGetDeviceInterfaces(guid, NULL, DEVICE_INTERFACE_INCLUDE_NONACTIVE, &SymbolicLinkList);
 
-    ok(Status == STATUS_INVALID_DEVICE_REQUEST,
-        "IoRegisterDeviceInterface returned 0x%08lX\n", Status);
+    ok(NT_SUCCESS(Status), "IoGetDeviceInterfaces failed with status 0x%X for '%wZ'\n", Status, &GuidString);
+    if (!NT_SUCCESS(Status))
```

Comment:
You can use `if (skip(NT_SUCCESS(Status), "No device interface list\n"))` to issue a skip message in this case. That helps make it more obvious that parts of the test couldn't be executed.

Reply:
OK, added.